contracts/concert/src/lib.rs

```diff
@@ -125,108 +136,102 @@ impl Concert {
             0,
         )
-        .expect("Error in async message to MTK contract")
+        .expect("Error in async message to Mtk contract")
         .await
         .expect("CONCERT: Error minting concert tokens");
 
-        msg::reply(
-            ConcertEvent::Purchase {
-                concert_id: self.concert_id,
-                amount,
-            },
-            0,
-        )
-        .expect("Error during a replying with ConcertEvent::Purchase");
+        Ok(ConcertEvent::Purchase {
+            concert_id: self.concert_id,
+            amount,
+        })
     }
 
     // MINT SEVERAL FOR A USER
-    async fn hold_concert(&mut self) {
+    async fn hold_concert(&mut self) -> Result<ConcertEvent, ConcertError> {
         if msg::source() != self.creator {
-            panic!("CONCERT: Only creator can hold a concert");
+            return Err(ConcertError::NotCreator);
         }
         // get balances from a contract
         let accounts: Vec<_> = self.buyers.clone().into_iter().collect();
         let tokens: Vec<TokenId> = iter::repeat(self.ticket_ft_id)
             .take(accounts.len())
             .collect();
 
-        let balance_response: MTKEvent = msg::send_for_reply_as(
+        let balance_response: MtkEvent = msg::send_for_reply_as(
             self.contract_id,
-            MyMTKAction::BalanceOfBatch {
+            MtkAction::BalanceOfBatch {
                 accounts,
                 ids: tokens,
             },
             0,
             0,
         )
-        .expect("Error in async message to MTK contract")
+        .expect("Error in async message to Mtk contract")
         .await
         .expect("CONCERT: Error getting balances from the contract");
         let balances: Vec<BalanceReply> =
-            if let MTKEvent::BalanceOf(balance_response) = balance_response {
+            if let MtkEvent::BalanceOf(balance_response) = balance_response {
                 balance_response
             } else {
                 Vec::new()
             };
         // we know each user balance now
         for balance in &balances {
-            msg::send_for_reply_as::<_, MTKEvent>(
+            msg::send_for_reply_as::<_, MtkEvent>(
                 self.contract_id,
-                MyMTKAction::Burn {
+                MtkAction::Burn {
                     id: balance.id,
                     amount: balance.amount,
                 },
                 0,
                 0,
             )
-            .expect("Error in async message to MTK contract")
+            .expect("Error in async message to Mtk contract")
             .await
             .expect("CONCERT: Error burning balances");
         }
 
         for actor in &self.buyers {
-            let mut ids = vec![];
-            let mut amounts = vec![];
-            let mut meta = vec![];
             let actor_metadata = self.metadata.get(actor);
             if let Some(actor_md) = actor_metadata.cloned() {
+                let mut ids = vec![];
```

Comment:
It is better to set capacity if known.
```suggestion
                let mut ids = Vec::with_capacity(actor_md.len());
```